keps/sig-cloud-provider/20191004-out-of-tree-credential-providers.md

```diff
@@ -0,0 +1,216 @@
+* Remove in-tree credential provider code from kuberentes core
+
+### Non-Goals
+
+* Broad removal of cloud SDK usage falls under the [KEP for removing in-tree providers](https://github.com/kubernetes/enhancements/blob/master/keps/sig-cloud-provider/2019-01-25-removing-in-tree-providers.md).
+* Continuing to support projects that import the credential provider package.
+
+## Proposal
+
+### External Credential Provider
+
+An executable capable of providing container registry credentials will be pre-installed on each node so that it exists when kubelet starts running. This binary will be executed by the kubelet to obtain container registry credentials in a format compatible with container runtimes. Credential responses may be cached within the kubelet.
+
+This architecture is similar to the approach taken by the exec based credential plugin architecture already present in client-go and CNI, and is a well understood pattern.  The API types are modeled after the ExecConfig and ExecCredential in client-go which define exec based credential retrieval for similar use cases.
+
+A RegistryCredentialProvider configuration API type (similar to [clientauthentication](https://github.com/kubernetes/kubernetes/tree/0273d43ae9486e9d0be292c01de2dd4143522b86/staging/src/k8s.io/client-go/pkg/apis/clientauthentication/v1beta1)) will be added to Kubernetes:
+
+```go
+// RegistryCredentialProvider is used by the kubelet container runtime to match image URLs with
+// exec-based credential provider plugins that provide container registry credentials.
+type RegistryCredentialProvider struct {
+    metav1.TypeMeta `json:",inline"`
+
+    // ImageURLMatchers is a list of strings used to match against the image URL
+    // to determine which images to provide credentials for.  If one of the
+    // string matches the image URL, then the RegistryCredentialProvider will be
+    // used by kubelet to provide credentials for the image pull.
+    //
+    // Each ImageURLMatcher string is a URL pattern which can optionally contain
+    // a port and a path.  Globs can be used in the hostname (but not the port or
+    // the path).
+    //
+    // An image URL matches when it has the same number of parts as the
+    // ImageURLMatcher string, and each part matches.  Additionally the path of
+    // ImageURLMatcher must be a prefix of the target URL. If the ImageURLMatch
+    // contains a port, then the port must match as well.
+    ImageURLMatch []string `json:"imageURLMatch,omitempty"`
+
+    // Exec specifies a custom exec-based plugin.  This type is defined in
+    // https://github.com/kubernetes/client-go/blob/62f256057db7571c5ed1aba47eea291f72dd557a/tools/clientcmd/api/types.go#L184
+    Exec clientcmd.ExecConfig
+}
+```
+
+The RegistryCredentialProvider will be encoded in YAML and located in a file on disk.  The directory where credential provider configuration is found will be passed to kubelet via a new configuration option `RegistryCredentialProviderDir`.
```

Comment:
if the kubelet is given a directory, and the config is in a file, how does the kubelet locate the config file? scan all yaml files? look for a specific filename? we've typically pointed the kubelet at the specific config file we want it to consume.

Reply:
Good point.  Pointing at a specific file (or files, if we support multiple) seems like the easiest way to implement this.

Reply:
It probably would be simplest to accept a single file with a format like:

```go
type RegistryCredentialConfig struct {
    metav1.TypeMeta `json:",inline"`

    Providers []RegistryCredentialProvider
}

type RegistryCredentialProvider struct {
    ImageURLMatchers []string
    Exec clientcmd.ExecConfig
}
```

Reply:
@enj good idea :+1: